### PR TITLE
[WIP] Improve my Scalafix: RFC

### PR DESCRIPTION
**Pull request by sideeffffect**
This adds many of the Scalazzi rules. (also sorts imports, but that's a small thing.)
What rules do we want to have. Which ones would we like to keep? Which ones would we like to add?
We would need to translate some rules from `scalaz` to `cats`.
Eventually we could migrate away from Wartremover.

Comments and suggestions welcome.

For example, we probably don't want to enforce `noDefaultArgs`, because default args is essential feature of config classes with PureConfig.
On the other hand, we might want to enforce `noValInAbstract`, because it could lead to initialization-related runtime bugs.

EDIT: some of the changes are only of whitespace, so I recommend to turn on _Hide whitespace changes_

### CODE REVIEW

File: .scalafix.conf [hunk 1]
```diff
@@ -6,16 +6,86 @@ rules = [
...
+  MissingFinal
 ]
 
 ExplicitResultTypes {
   unsafeShortenNames = true
 }
 
+SortImports.blocks = [
+  "java."
+  "*"
```
Comment (hanny24) [inline .scalafix.conf]: Suggestion: maybe treat com.avast as separate block

File: project/Dependencies.scala [hunk 2]
```diff
@@ -27,6 +27,7 @@ object Dependencies {
   val pureConfig = "com.github.pureconfig" %% "pureconfig" % "0.12.2"
   val scalaTest = "org.scalatest" %% "scalatest" % "3.1.0"
   val scalazzi = "com.github.vovapolu" %% "scaluzzi" % "0.1.3"
+  val sortImports = "com.nequissimus" %% "sort-imports" % "0.3.2"
```
Comment (hanny24) [inline project/Dependencies.scala]: Does it work reliably? Last time I tried it it removed some imports that were essential

File: cassandra-datastax-driver/src/main/scala/com/avast/sst/datastax/DatastaxHelper.scala [hunk 3]
```diff
@@ -1,9 +1,9 @@
...
 
-import com.datastax.oss.driver.api.core.config.{DriverOption, ProgrammaticDriverConfigLoaderBuilder => DriverBuilder}
-
-import scala.collection.JavaConverters._
-import scala.concurrent.duration.Duration
+import com.datastax.oss.driver.api.core.config.{DriverOption, ProgrammaticDriverConfigLoaderBuilder => DriverBuilder}
+
+

import scala.collection.JavaConverters._
```
Comment (hanny24) [inline cassandra-datastax-driver/src/main/scala/com/avast/sst/datastax/DatastaxHelper.scala]: ??

File: .scalafix.conf [hunk 4]
```diff
@@ -34,3 +104,127 @@ Disable.ifSynthetic = [
...
+//        }
+//        message = "Deterministic method is not total, must be called via Maybe.attempt, \/.fromTryCatchNonFatal, etc."
+//      }
+//    ]
+//  }
+//]
+
+DisableSyntax {
+  noAsInstanceOf = true
+  noContravariantTypes = true
```
Comment (hanny24) [inline .scalafix.conf]: This seems too much.

File: .scalafix.conf [hunk 5]
```diff
@@ -34,3 +104,127 @@ Disable.ifSynthetic = [
...
+  noCovariantTypes = true
```
Comment (hanny24) [inline .scalafix.conf]: This seems too much

File: .scalafix.conf [hunk 6]
```diff
@@ -34,3 +104,127 @@ Disable.ifSynthetic = [
...
+//  noDefaultArgs = true
+  noFinalVal = true
+  noFinalize = true
+  noImplicitConversion = true
+  noImplicitObject = true
+  noIsInstanceOf = true
+  noNulls = true
```
Comment (hanny24) [inline .scalafix.conf]: It might make integration with java libs very difficult.

File: .scalafix.conf [hunk 7]
```diff
@@ -34,3 +104,127 @@ Disable.ifSynthetic = [
...
+  noReturns = true
+  noSemicolons = true
+  noTabs = true
+  noThrows = true
+  noUniversalEquality = true
+//  noValInAbstract = true
+  noValPatterns = true
```
Comment (hanny24) [inline .scalafix.conf]: Probably not in tests

File: .scalafix.conf [hunk 8]
```diff
@@ -34,3 +104,127 @@ Disable.ifSynthetic = [
...
+  noVars = true
```
Comment (hanny24) [inline .scalafix.conf]: Again, seems to harsh.

File: .scalafix.conf (hunk 1 above)
Comment (sideeffffect) [reply .scalafix.conf]: :thinking:

File: project/Dependencies.scala (hunk 2 above)
Comment (sideeffffect) [reply project/Dependencies.scala]: has been working reliably for me so far :man_shrugging:

File: cassandra-datastax-driver/src/main/scala/com/avast/sst/datastax/DatastaxHelper.scala (hunk 3 above)
Comment (sideeffffect) [reply cassandra-datastax-driver/src/main/scala/com/avast/sst/datastax/DatastaxHelper.scala]: problems with LF vs CRLF, I've fixed it already

File: .scalafix.conf (hunk 6 above)
Comment (sideeffffect) [reply .scalafix.conf]: do you think it would be _too much_ even if it can be suppressed with an annotation?

**Comment by sideeffffect** [comment]
> disable specific check using comment
yep. with annotation, to be precise. see the diff, I've already did it at some places

File: .scalafix.conf (hunk 6 above)
Comment (hanny24) [reply .scalafix.conf]: In that case it's ok

**Comment by jakubjanecek** [comment]
It's a lot of rules at once. I would like to understand what we sign up for. So even though I understand it is a lot of work I would like to see a list of rules that are added with some short explanation why we want to add them (unless it's really obvious). For example the thing with covariant and contravariant types makes no sense to me. Disabling `java.io` also seems very strange - I know it's blocking but blocking does not need to be bad!

**Comment by jakubjanecek** [comment]
Also... what is the relation of import sorting between Scalafmt and Scalafix?

Also, please remove the commented parts - they shouldn't be there and make the PR bigger and more confusing. Thanks.

**Comment by sideeffffect** [comment]
> what is the relation of import sorting between Scalafmt and Scalafix

they complement each other. Scalafmt sorts imports that are on one line (ala `import x.{y, z}`). Scalafix is sorting the import lines.

File: .scalafix.conf (hunk 7 above)
Comment (sideeffffect) [reply .scalafix.conf]: unfortunately I don't know to enforce a rule only in `Compile`, and ignore it in `Test` :slightly_frowning_face:

File: .scalafix.conf (hunk 8 above)
Comment (sideeffffect) [reply .scalafix.conf]: hmm, I think that in the rare cases when you need `var` and know what you're doing, you can suppress this rule at the place

**Comment by sideeffffect** [comment]
@jakubjanecek I open the PR even with commented rules, so that we can have a discussion about them -- otherwise how could we decide if we want them or not.

I personally don't have a clear idea about what rules to enforce and which would be counterproductive.

Some of the rules are described here: https://github.com/scalacenter/scalafix/blob/master/scalafix-rules/src/main/scala/scalafix/internal/rule/DisableSyntaxConfig.scala#L18

Btw, maybe I should have mentioned it in the beginning -- this scalafix config comes from [`scalazzi`](https://github.com/scalaz/scalazzi/blob/master/scalafix.conf) and in this PR I would like us to discuss it and transform it into something suiting us.
Regarding variance, we all agree, so I've commented it out. Regarding `java.io`, can you please put a comment under the rule?

**Comment by jakubjanecek** [comment]
This PR was split into many smaller PRs which were accepted with some changes.

**Pull request closed without merging**